### ISSUE TITLE
[Bug] Lexer lookup doesn't cache custom plugin lexers

### ISSUE THREAD

**Issue by kamilkrzyskow** [pygments/pygments#2479]
Hi 👋,
I'm coming from https://github.com/facelessuser/pymdown-extensions/pull/2121 where I tried to add a simple cache for the lexer lookup to make my MkDocs documentation build time faster. During the process it turned out that the issue is in Pygments, because the module does implement a cache methodology for built-in lexers, but doesn't do so for custom lexers from the entry points.

I'm specifically talking about this function, which is used by Python Markdown Extensions: 
https://github.com/pygments/pygments/blob/master/pygments/lexers/__init__.py#L109-L131/

Since I'm using MkDocs with Python Markdown Extensions, my usage of Pygments is rather streamlined to providing the appropriate markdown code block with the custom lexer alias name `dae` and this is everywhere throughout the documentation. Therefore, I would prefer for the cache to incorporate that and make use of the alias itself when searching the cache... What I'm alluding to is that the cache lookup should be before the `for` loop to keep it O(1) and not O(N) + O(1)

In this comment https://github.com/facelessuser/pymdown-extensions/pull/2121#issuecomment-1646689381 I provided my testing methodology, which has shown the difference in time of the cache lookup of different builtin lexers. It also shows that the time of lookup for the custom lexer stays the same...

**Comment by kamilkrzyskow** [comment]
Here are some diagnostics files, but as you added in the second message this affects everyone that has a lot of code examples in their documentation, which use a custom lexer.
[pip_freeze.txt](https://github.com/pygments/pygments/files/12138541/pip_freeze.txt)
[DxDiag.txt](https://github.com/pygments/pygments/files/12138542/DxDiag.txt)
```
>>> platform.python_version()
'3.11.3'
```

The lexer plugin is open source: https://github.com/kamilkrzyskow/gothic-lexer  

On my machine the custom lexer lookup is ~20ms, which I'm aware is exaggerated by my slow PC. However, the issue of the slow documentation build time was noticeable even before, on my Ryzen 3600 machine, because I turned off the Pygment highlighting for the development build. It wasn't 33 -> 4 seconds (explained [here](https://github.com/facelessuser/pymdown-extensions/pull/2121#issue-1816804956)), but something along the lines of 13 -> 3 seconds. I'm currently away, and can't use Ryzen 3600 machine to compare.

To summarize my previous post, since I see there are too many words:
The issues are here twofold, the cache doesn't affect custom lexers, the cache lookup comes after the lexer lookup instead of before it.
